hyx/ratelimit/buckets.py

```diff
@@ -0,0 +1,66 @@
+import asyncio
+from typing import Optional
+
+from hyx.ratelimit.exceptions import RateLimitExceeded
+
+
+class TokenBucket:
+    """
+    Token Bucket Logic
+    Replenish tokens as time passes on. If tokens are available, executions can be allowed.
+    Otherwise, it's going to be rejected with RateLimitExceeded
+    """
+
+    __slots__ = (
+        "_max_executions",
+        "_per_time_secs",
+        "_bucket_size",
+        "_loop",
+        "_token_per_secs",
+        "_tokens",
+        "_next_replenish_at",
+    )
+
+    def __init__(self, max_executions: float, per_time_secs: float, bucket_size: Optional[float] = None) -> None:
+        self._max_executions = max_executions
+        self._per_time_secs = per_time_secs
+
+        self._bucket_size = bucket_size if bucket_size else max_executions
+
+        self._loop = asyncio.get_running_loop()
+        self._token_per_secs = self._per_time_secs / self._max_executions
+
+        self._tokens = self._bucket_size
+        self._next_replenish_at = self._loop.time() + self._token_per_secs
+
+    @property
+    def tokens(self) -> float:
+        return self._tokens
+
+    @property
+    def empty(self) -> bool:
+        return self._tokens <= 0
+
+    async def acquire(self) -> None:
+        if not self.empty:
+            self._tokens -= 1
+            return
+
+        now = self._loop.time()
+
+        next_replenish = self._next_replenish_at
```

Comment:
In the `TokenBucketRateLimiter` case, this used to be the only place where we needed to check for bucket replenishment. 
But when you shift that view to a `TokenBucket` class which is an independent component, meant to be used elsewhere, it makes a lot of sense to check it in more places.

For example, if we did not check for replenishment in `tokens()` and `empty()` methods before returning their values that are based on the number of tokens, we would return outdated values (b/c there is no assumption how the usage patterns of the TokenBucket component look).

That's why it makes sense to move the replenishment logic to a separate private method and then use it in all three methods.

Reply:
A little confused about what this separate private method is going to contain, specifically when being called in the `empty()` and `tokens()` methods. In acquire(), I understand that the method first checks if the token bucket is empty. If it is, it will check if a certain amount of time has passed for it to be able to continue to update the number of tokens while also updating the next time to replenish value. Should this separate private method only contain the check or also the update token and time portion of this method?

If it only contains the checK:
- how would this ensure that values aren't outdated?
- Where would perform the update?

If the separate method contains both the check and update: 
- Should both methods empty() and tokens() check if the token bucket is empty, and if we do need to check, what would be the point of adding the separate method in the `empty()` method as the `empty()` method would never return true after the number of tokens has been updated?

Reply:
@kevkevy3000 here is a test case I have been thinking of. It would fail to behave as expected with the current way we update the bucket.

Pseudocode:

```python
bucket = TokenBucket("3 tokens", "replenish 1 token per sec")

bucket.take("3 tokens") # take 3 tokens out of the bucket, 0 tokens left

# at this point we have 0 tokens in the bucket
sleep(3 seconds)

# at this point we should have 3 tokens in the bucket because we have just waited 3 secs and each sec we should one token

bucket.tokens  # should be 3 tokens, but we would get 0
bucket.empty # should be False, but we would get True
```

Based on this, I hope it's clear that the `_replenish()` method should try to replenish the bucket if possible e.g. check the last time we updated the bucket and try to add some tokens to the bucket. Otherwise, the token state would be inconsistent with respect to the time that passed.

So essentially the `_replenish()` method should contains this:

> it will check if a certain amount of time has passed for it to be able to continue to update the number of tokens while also updating the next time to replenish value.

The new `take()`/`acquire()` method logic should be no different to `tokens`/`empty` properties and it should also start from trying to replenish the bucket, and only then proceed trying to take a given number of tokens off the bucket.

Hope this makes more sense.